package com.ddd4j.cloud.core.contract;

import com.ddd4j.cloud.core.contract.exception.ServiceException;

import java.io.Serializable;

/**
 * 统一接口响应，标准的响应数据结构
 *
 * @author Jensen
 * @公众号 架构师修行录
 */
public interface IR extends Serializable {
    Serializable getCode();

    String getMsg();

    <T> T getData();

    Boolean isOk();

    default void isOk(String notOkThrows) {
        if (!isOk()) {
            throw new ServiceException(notOkThrows + " -> {}", this);
        }
    }

    default <T> T getData(String notOkThrows) {
        if (!isOk()) {
            throw new ServiceException(notOkThrows + " -> {}", this);
        }
        T data = getData();
        if (data == null) {
            throw new ServiceException(notOkThrows + " -> {}", this);
        }
        return data;
    }
}